### PR TITLE
feat: enable data availiability verification via vote extensions

### PR DESCRIPTION
**Pull request by Teja2045**
ref : #24 #16

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package client
+
+import (
```
Comment (PrathyushaLakkireddy) [inline client/client.go:3]: move this file to chainclient dir

File: keeper/abci.go [hunk 2]
```diff
@@ -1,80 +1,187 @@
...
+	"encoding/json"
+	"fmt"
+
 	abci "github.com/cometbft/cometbft/abci/types"
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	"github.com/vitwit/avail-da-module/types"
 )
 
+type StakeWeightedVotes struct {
```
Comment (PrathyushaLakkireddy) [inline keeper/abci.go:13]: for this type of structs,let's maintain a seperate file(types.go) and move these structs to there.

File: keeper/abci.go [hunk 3]
```diff
@@ -1,80 +1,187 @@
...
+			from := k.GetStartHeightFromStore(ctx)
+			to := k.GetEndHeightFromStore(ctx)
+
+			pendingRangeKey := Key(from, to)
+			votingPower := injectedVoteExtTx.Votes[pendingRangeKey]
+
+			if votingPower > 0 { // TODO: calculate voting power properly
+				k.setBlobStatusSuccess(ctx)
+			} else {
+				k.SetBlobStatusFailure(ctx)
```
Comment (PrathyushaLakkireddy) [inline keeper/abci.go:102]: Combine these two functions, into one.

File: keeper/keeper.go [hunk 4]
```diff
@@ -23,28 +29,32 @@ type Keeper struct {
...
 	storeKey storetypes2.StoreKey
 
 	cdc codec.BinaryCodec
 
 	publishToAvailBlockInterval int
+	PublishToAvailBlockInterval uint64
+	MaxBlocksForBlob            uint
+	VotingInterval              uint64
 	injectedProofsLimit         int
 	app_id                      int
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:46]:
```suggestion
	appID                      int
```

File: keeper/keeper.go [hunk 5]
```diff
@@ -23,28 +29,32 @@ type Keeper struct {
...
+	ProvenHeight            collections.Item[uint64]
 	PendingBlocksToTimeouts collections.Map[int64, int64]
 	TimeoutsToPendingBlocks collections.Map[int64, types.PendingBlocks]
+	keyring                 keyring.Keyring
 
 	storeKey storetypes2.StoreKey
 
 	cdc codec.BinaryCodec
 
 	publishToAvailBlockInterval int
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:41]:
```suggestion
```

File: keeper/keeper.go [hunk 6]
```diff
@@ -23,28 +29,32 @@ type Keeper struct {
...
+	PublishToAvailBlockInterval uint64
+	MaxBlocksForBlob            uint
+	VotingInterval              uint64
 	injectedProofsLimit         int
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:45]:
```suggestion
```

File: keeper/keeper.go [hunk 7]
```diff
@@ -70,10 +79,72 @@ func NewKeeper(
...
+func (k *Keeper) GetBlobStatus(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+	return GetStatusFromStore(store)
+}
+
+// Todo: remove this method later
+func (k *Keeper) SubmitBlob(ctx sdk.Context, req *types.MsgSubmitBlobRequest) (*types.MsgSubmitBlobResponse, error) {
+
+	return &types.MsgSubmitBlobResponse{}, nil
+}
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:102]:
```suggestion

```

File: keeper/msg_server.go [hunk 8]
```diff
@@ -37,3 +39,24 @@ func (s msgServer) SetAvailAddress(ctx context.Context, msg *types.MsgSetAvailAd
...
+	//TODO: query the light client
+	return s.k.UpdateBlobStatus(sdkCtx, req)
+}
+
+/*
+rpc SubmitBlob(MsgSubmitBlobRequest) returns (MsgSubmitBlobResponse);
+
+  // UpdateBlobStatus
+  rpc UpdateBlobStatus(MsgUpdateBlobStatusRequest) returns (MsgUpdateBlobStatusResponse);
+*/
```
Comment (PrathyushaLakkireddy) [inline keeper/msg_server.go:62]:
```suggestion
```

File: keeper/abci.go [hunk 9]
```diff
@@ -1,80 +1,184 @@
...
+	var blocksToSumit []int64
 
-	if err := k.preblockerPendingBlocks(ctx, req.Time, req.ProposerAddress, &injectedData.PendingBlocks); err != nil {
-		return err
+	for i := fromHeight; i < endHeight; i++ {
+		blocksToSumit = append(blocksToSumit, int64(i))
 	}
-	// }
+
+	// only proposar should should run the this
```
Comment (PrathyushaLakkireddy) [inline keeper/abci.go:125]:
```suggestion
	// only the proposer should be able to post the blocks
```

File: keeper/abci.go [hunk 10]
```diff
@@ -1,80 +1,184 @@
...
+			h.VoteExtHandler.logger.Error("failed to decode vote extension", "err", err, "validator", fmt.Sprintf("%x", v.Validator.Address))
+			continue
 		}
+
+		if voteExt.Votes == nil {
+			continue
+		}
+
+		// TODO: remove if this is not used anywhere
+		totalStake += v.Validator.Power
```
Comment (PrathyushaLakkireddy) [inline keeper/abci.go:172]:
```suggestion
		
```

File: keeper/abci.go [hunk 11]
```diff
@@ -1,80 +1,184 @@
...
+}
+
+func (h *ProofOfBlobProposalHandler) aggregateVotes(ctx sdk.Context, ci abci.ExtendedCommitInfo) (map[string]int64, error) {
+	from := h.keeper.GetStartHeightFromStore(ctx)
+	to := h.keeper.GetEndHeightFromStore(ctx)
+
+	pendingRangeKey := Key(from, to)
+	votes := make(map[string]int64, 1)
+
+	var totalStake int64
```
Comment (PrathyushaLakkireddy) [inline keeper/abci.go:152]:
```suggestion
```

File: keeper/keeper.go [hunk 12]
```diff
@@ -54,26 +62,80 @@ func NewKeeper(
...
 
 		cdc: cdc,
 
-		publishToAvailBlockInterval: publishToAvailBlockInterval,
-		app_id:                      appId,
+		appID: appId,
 
-		unprovenBlocks: make(map[int64][]byte),
+		unprovenBlocks:              make(map[int64][]byte),
+		MaxBlocksForBlob:            20, //Todo: call this from app.go, later change to params
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:80]: For now, I think we can read the value from config itself(just like the way we added avail config in our previous approach). Wdyt.

File: keeper/keeper.go [hunk 13]
```diff
@@ -54,26 +62,80 @@ func NewKeeper(
...
+	endHeight := k.GetEndHeightFromStore(ctx)
+	status := GetStatusFromStore(store)
+
+	if req.BlocksRange.From != provenHeight+1 || req.BlocksRange.To != endHeight {
+		return nil, fmt.Errorf("invalid blocks range request: expected range [%d -> %d], got [%d -> %d]",
+			provenHeight+1, endHeight, req.BlocksRange.From, req.BlocksRange.To)
+	}
+
+	if status != PENDING_STATE {
+		return nil, errors.New("can update the status if it is not pending")
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:108]:
```suggestion
		return nil, errors.New("can't update the status if it is not pending")
```

File: keeper/keeper.go [hunk 14]
```diff
@@ -54,26 +62,80 @@ func NewKeeper(
...
+		UpdateAvailHeight(ctx, store, req.AvailHeight)                            // updates avail height at which the blocks got submitted to DA
+		UpdateVotingEndHeight(ctx, store, uint64(currentHeight)+k.VotingInterval) // TODO: Now voting interval is 5, so check whether we can process votes at next block after tx exec.
+	}
+
+	UpdateBlobStatus(ctx, store, newStatus) // updates blob status after based on tx exec
+
+	return &types.MsgUpdateBlobStatusResponse{}, nil
+}
+
+func (k *Keeper) SubmitBlobStatus(ctx sdk.Context, _ *types.QuerySubmitBlobStatusRequest) (*types.QuerySubmitBlobStatusResponse, error) {
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:125]:
```suggestion
func (k *Keeper) SubmittedBlobStatus(ctx sdk.Context, _ *types.QuerySubmittedBlobStatusRequest) (*types.QuerySubmittedBlobStatusResponse, error) {
```

Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:125]: I think we can move this methods to grpc_query server and msg_server files respectively

File: keeper/keeper.go [hunk 15]
```diff
@@ -54,26 +62,80 @@ func NewKeeper(
...
+	// Todo: implement query
+	store := ctx.KVStore(k.storeKey)
+	startHeight := k.GetStartHeightFromStore(ctx)
+	endHeight := k.GetEndHeightFromStore(ctx)
+	status := GetStatusFromStore(store)
+	statusString := ParseStatus(status)
```
Comment (PrathyushaLakkireddy) [inline keeper/keeper.go:131]:
```suggestion
	blobStatus := ParseStatus(status)
```

File: relayer/client.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/cosmos/go-bip39"
+
+	// "github.com/tendermint/starport/starport/pkg/xfilepath"
+
+	"github.com/cosmos/cosmos-sdk/types/module"
+)
+
+const (
+	defaultGasAdjustment = 1.0
+	defaultGasLimit      = 300000
```
Comment (PrathyushaLakkireddy) [inline relayer/client.go:28]:
```suggestion
```

File: relayer/publish.go [hunk 17]
```diff
@@ -49,30 +49,83 @@ func (r *Relayer) postBlocks(ctx sdk.Context, blocks []int64) {
...
+		}, cdc)
+		if err != nil {
+			fmt.Println("error while submitting tx...", err)
+		}
+
+		return
+	}
+
+	if blockInfo.BlockNumber != 0 {
+		fmt.Println("proposer addressss........", sdk.AccAddress.String(proposer),
```
Comment (PrathyushaLakkireddy) [inline relayer/publish.go:112]:
```suggestion
```

File: relayer/submit_data.go [hunk 18]
```diff
@@ -14,36 +12,32 @@ import (
...
+func (r *Relayer) SubmitDataToAvailClient(Seed string, AppID int, data []byte, blocks []int64, lightClientUrl string) (BlockInfo, error) {
+	var blockInfo BlockInfo
+	// if r.submittedBlocksCache[blocks[0]] {
+	// 	return blockInfo, nil
+	// }
 
-	r.submittedBlocksCache[blocks[0]] = true
-	delete(r.submittedBlocksCache, blocks[0]-int64(len(blocks)))
+	// // r.submittedBlocksCache[blocks[0]] = true
+	// delete(r.submittedBlocksCache, blocks[0]-int64(len(blocks)))
```
Comment (PrathyushaLakkireddy) [inline relayer/submit_data.go:25]: if we are not using it remove these line.

**Review by PrathyushaLakkireddy — APPROVED** [review]

**Review by NagaTulasi — APPROVED** [review]

**Pull request merged**